src/nlp_expr.jl

```diff
@@ -0,0 +1,562 @@
+            # We can use recursion here, because NonlinearExpr only occur in
+            # other NonlinearExpr.
+            MOI.Nonlinear.parse_expression(data, expr, arg, parent_node)
+        end
+    end
+    return
+end
+
+function _get_node_type(data, x)
+    id = get(data.operators.univariate_operator_to_id, x.head, nothing)
+    if length(x) == 1 && id !== nothing
+        return id, MOI.Nonlinear.NODE_CALL_UNIVARIATE
+    end
+    id = get(data.operators.multivariate_operator_to_id, x.head, nothing)
+    if id !== nothing
+        return id, MOI.Nonlinear.NODE_CALL_MULTIVARIATE
+    end
+    id = get(data.operators.comparison_operator_to_id, x.head, nothing)
+    if id !== nothing
+        return id, MOI.Nonlinear.NODE_COMPARISON
+    end
+    id = get(data.operators.logic_operator_to_id, x.head, nothing)
+    if id !== nothing
+        return id, MOI.Nonlinear.NODE_LOGIC
+    end
+    return throw(MOI.UnsupportedNonlinearOperator(x.head))
+end
+
+function _parse_without_recursion_inner(stack, data, expr, x, parent)
+    id, node_type = _get_node_type(data, x)
+    push!(expr.nodes, MOI.Nonlinear.Node(node_type, id, parent))
+    parent = length(expr.nodes)
+    for i in length(x):-1:1  # Args need to be pushed onto the stack in reverse
+        push!(stack, (parent, x[i]))
+    end
+    return
+end
+
+# Method definitions
+
+Base.zero(::Type{NonlinearExpr}) = NonlinearExpr(:+, 0.0)
+
+Base.one(::Type{NonlinearExpr}) = NonlinearExpr(:+, 1.0)
+
+# Univariate operators
+
+for f in MOI.Nonlinear.DEFAULT_UNIVARIATE_OPERATORS
+    op = Meta.quot(f)
+    if f == :+
+        continue  # We don't need this.
+    elseif f == :-
+        @eval Base.:-(x::NonlinearExpr) = NonlinearExpr(:-, x)
+    elseif isdefined(Base, f)
+        @eval Base.$(f)(x::AbstractJuMPScalar) = NonlinearExpr($op, x)
+    elseif isdefined(MOI.Nonlinear, :SpecialFunctions)
+        # The operator is defined in some other package.
+        SF = MOI.Nonlinear.SpecialFunctions
+        if isdefined(SF, f)
+            @eval $(SF).$(f)(x::AbstractJuMPScalar) = NonlinearExpr($op, x)
+        end
+    end
+end
+
+# Multivariate operators
+
+# The multivariate operators in MOI are +, -, *, ^, /, ifelse, atan
+#
+# However, ifelse is a builtin, so we can't add methods to it.
+
+# We need only very generic fallbacks for these, because all other cases are
+# caught with more specific methods.
+for f in (:+, :-, :*, :^, :/, :atan)
+    op = Meta.quot(f)
+    @eval begin
+        function Base.$(f)(x::AbstractJuMPScalar, y::_Constant)
+            rhs = convert(Float64, _constant_to_number(y))
+            return NonlinearExpr($op, x, rhs)
+        end
+        function Base.$(f)(x::_Constant, y::AbstractJuMPScalar)
+            lhs = convert(Float64, _constant_to_number(x))
+            return NonlinearExpr($op, lhs, y)
+        end
+        function Base.$(f)(x::AbstractJuMPScalar, y::AbstractJuMPScalar)
+            return NonlinearExpr($op, x, y)
+        end
+    end
+end
```

Comment:
@pulsipher I think one fix is that `+(::NonlinearExpr, ::NonlinearExpr)` could be intercepted to check if one of the terms was already a `:+` node.

Reply:
Done